core/commands/dag/dag.go

```diff
@@ -307,6 +307,24 @@ The output of blocks happens in strict DAG-traversal, first-seen, order.
 		// if err := car.Write(pipeW); err != nil {}
 
 		pipeR, pipeW := io.Pipe()
+		// this is so convoluted because we:
+		// - want to close the readpipe on error so that the goroutine dies
+		// - can not do so on success, because during a no-deamon process
+		//   scenario the PostRun will not be able to grab the EOF from the
+		//   now-closed pipe
+		//
+		// Skipping this check results in:
+		//
+		// 		cmd/ipfs/ipfs dag export QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm >/dev/null
+		//  0s  0 B / ? [------------------------------------------------------------------------=]Error: io: read/write on closed pipe
+		//  1s  107.08 MiB / ? [-----------------------------------------=----------] 106.59 MiB/s
+		//
+		var emittingError error
+		defer func() {
+			if emittingError != nil {
+				pipeR.Close()
+			}
+		}()
```

Comment:
Given all this complexity, I'd just move this back to where it was.

Reply:
There are 3 spots where we return errors. I would need to add a closer to all of them... Do that?

Reply:
We only need it in https://github.com/ipfs/go-ipfs/pull/7036/files/63a56cf7bee571c5c1408a198e68be9f197bc6a7..70834450926f53902039ad3fbcc62c07b6f37758#diff-15172926c4147422f472139313467179R352. The next line waits on the error channel (which won't return till we're done writing so there's no need to unblock the writer).